### PR TITLE
PHPLIB-954: Add return types to all methods

### PR DESCRIPTION
**Pull request by alcaeus**
PHPLIB-954

This PR:
* adds testing with the 2.0 branch for PHPC,
* adds native return types for all methods, removing `ReturnTypeWillChange` attributes where present
* replaces all `Cursor` types with `CursorInterface`
* removes the compatibility layer introduced for the `CursorId` deprecation

### CODE REVIEW

File: src/BulkWriteResult.php
Comment (alcaeus) [inline src/BulkWriteResult.php:43]: These methods can be refactored when PHPC-2144 is implemented. At that point, the return type of `WriteResult` methods will change to `int` and they will always throw when accessing any result counts for an unacknowledged write.

File: src/Collection.php
Comment (alcaeus) [inline src/Collection.php]: We don't have any plans for this yet, but I propose having `CursorInterface` extend `Iterator` to avoid the need for an intersection type here. Currently, only the `Cursor` class implements `Iterator` (added in PHPC-1691, but not added to `CursorInterface` due to BC constraints).

File: src/Operation/DropSearchIndex.php
Comment (alcaeus) [inline src/Operation/DropSearchIndex.php]: The void type hint is invalid, as the interface has a `mixed` type. Since `mixed` is a type that can't combined with other types, but also doesn't include `void`, we have to use `mixed` for now. We can change this when upgrading to PHP 8.2, as that version introduces `null` as a standalone type. If we were to make this class final, we could make the change in an upcoming minor version, especially as the method only ever returns `null`.

File: src/Operation/Explainable.php
Comment (alcaeus) [inline src/Operation/Explainable.php:31]: We currently return arrays for all commands - this major version would be a chance to create objects instead if we wanted to. I don't have any strong feelings about this, but wanted to mention it.

File: src/Operation/UpdateSearchIndex.php
Comment (alcaeus) [inline src/Operation/UpdateSearchIndex.php]: I decided to return the `Cursor` instance returned from `executeCommand` here. We could also return `null` as was done in `DropSearchIndex` and change the return type in an upcoming minor version.

**Review by alcaeus — COMMENTED** [review]
Left some explanations. A number of the psalm errors are due to https://github.com/vimeo/psalm/issues/10019, which doesn't correctly infer the return type or `(object) ['foo' => 'bar']` as `stdClass` but instead assumes `object`. I'll take a look at the other psalm errors regarding codec generics, but feel free to leave feedback in the meantime.

File: src/Client.php
Comment (GromNaN) [inline src/Client.php:146]: We could describe the return type in detail with the types of each field.

File: src/Collection.php
Comment (GromNaN) [reply src/Collection.php]: `CursorInterface` already extends `Traversable`, which is better IMO. We should not require the `Iterator` interface for custom cursor classes.
In case people really need the `Iterator` API, they can wrap it in an `IteratorIterator`.
```suggestion
    public function aggregate(array $pipeline, array $options = []): CursorInterface
```

File: src/Database.php
Comment (GromNaN) [inline src/Database.php]: I would expect a `CursorInterface` like `Collection::aggregate()`

File: src/Database.php
Comment (GromNaN) [inline src/Database.php]: Not a `CursorInterface`?
```suggestion
    public function command(array|object $command, array $options = []): CursorInterface
```

File: src/Operation/UpdateSearchIndex.php
Comment (GromNaN) [reply src/Operation/UpdateSearchIndex.php]: We discussed returning `void` for all non-meaningful results: [PHPLIB-1227](https://jira.mongodb.org/browse/PHPLIB-1227)

File: src/Database.php
Comment (alcaeus) [reply src/Database.php]: Good point, will update.

File: src/Operation/UpdateSearchIndex.php
Comment (alcaeus) [reply src/Operation/UpdateSearchIndex.php]: This won't work - best we can do is to return `null` as we can't change a return type to `void` if the interface specifies the method as returning a value.

File: src/Collection.php
Comment (alcaeus) [reply src/Collection.php]: A `CursorInterface` type without an intersection with `Iterator` won't allow you to call `$cursor->next();` - that's an issue we faced in ODM. Ideally, we'd fix this in PHPC by having `CursorInterface` extend `Iterator`, but until then we're stuck with this solution.

File: src/Database.php
Comment (alcaeus) [reply src/Database.php]: Updated

File: src/Operation/UpdateSearchIndex.php
Comment (GromNaN) [reply src/Operation/UpdateSearchIndex.php]: This interface is useless. It is never used to type, and in any case it is not capable of indicating the return type. At best, it guarantees a certain consistency between classes.

At the least, I'd remove the return type, but I'd be more inclined to remove this interface.

File: src/Client.php
Comment (alcaeus) [reply src/Client.php:146]: Technically, yes. However, since this method is only used when calling `var_dump` and we don't want users to calling it manually, I don't see the benefit of properly defining the array shape for static analysis.

File: src/Collection.php
Comment (GromNaN) [reply src/Collection.php]: There is always solutions to wrap a Transversable that is not an `Iterator`.
With a Generator (see [UnrewindableIterator](https://github.com/doctrine/mongodb-odm/blob/8e0deb186540a89111181c3be89924a3d791045c/lib/Doctrine/ODM/MongoDB/Iterator/UnrewindableIterator.php#L126-L127]), `IteratorIterator` or `IteratorAggregate::getIterator()`

Comment (jmikola) [reply src/Collection.php]: Providing some additional context and throwing in my two cents:

 * We added Iterator to Cursor in [PHPC-1691](https://jira.mongodb.org/browse/PHPC-1691) in order to comply with PHP 8.
 * CursorInterface was introduced in [PHPC-1123](https://jira.mongodb.org/browse/PHPC-1123) to allow other classes to fill in for a Cursor, which cannot be extended. AFAIK, the only existing use case for this within our own driver is CodecCursor.
 * CodecCursor currently implements both CursorInterface and Iterator, so that would not be negatively impacted by this change.

While conditionally wrapping a CursorInterface to get access to Iterator methods is always possible, it may be tedious. Given that CursorInterface is intended to be used for classes that compose a Cursor, it seems a small ask for the wrapping class to also implement Iterator.

I'm not sure what the use case would be for a CursorInterface implementation to rely on IteratorAggregate, as it would then have no way to influence the inner Cursor's iteration. Hypothetically, it could proxy other methods (e.g. `toArray()`), but I've never encountered such an implementation in the wild.

If having CursorInterface extend Iterator doesn't negatively impact PHPLIB, Doctrine ODM, and the Laravel integration, I wouldn't be too concerned about other use cases. Worst case, we're asking users to implement a implement a few extra methods (easily accomplished with a trait if needed).

File: src/Operation/DropSearchIndex.php
Comment (jmikola) [reply src/Operation/DropSearchIndex.php]: Would it be preferable to just remove the return type from `Executable::execute()` interface? This interface has always been internal and doesn't really serve any purpose within the library.

It is extended by Explainable, but that also seems unnecessary since the Explain operation only depends on `Explainable::getCommandDocument()`.

For context: Executable was created in a different time where I was adamant about everything having an interface ([PHPC-378](https://jira.mongodb.org/browse/PHPC-378)). Nine years later, I think we'd have done better to embrace the [YAGNI](https://en.wikipedia.org/wiki/You_aren't_gonna_need_it) philosophy earlier on.

File: src/Operation/Explainable.php
Comment (jmikola) [reply src/Operation/Explainable.php:31]: The pedantic choice would probably be `array|object` for consistency with [`MongoDB\Driver\Command`](https://www.php.net/manual/en/class.mongodb-driver-command.php).

**/me glances at the wall** ([yes, really](https://x.com/jmikola/status/1701971766137004529))

[Covariance](https://www.php.net/manual/en/language.oop5.variance.php) would still allow us to maintain `array` type hints in the operation class implementations: https://3v4l.org/AvFKa

And I don't think the Explain operation would require any changes one way or another.

File: src/Operation/UpdateSearchIndex.php
Comment (jmikola) [reply src/Operation/UpdateSearchIndex.php]: Saw this after replying in https://github.com/mongodb/mongo-php-library/pull/1391#discussion_r1759155384.

Removing the interface's return type hint is an easy solution. I also have no objections to removing the interface entirely, as it's internal and doesn't provide any real value. Operations are only ever used interchangeably via the Explainable interface (which really doesn't need to extend Executable).

File: src/Database.php
Comment (jmikola) [reply src/Database.php]: For context, when `Database::aggregate()` was implemented in https://github.com/mongodb/mongo-php-library/commit/c2922d30f2122c285b4a75db3b8b114b6e233350, the [Aggregate operation](https://github.com/mongodb/mongo-php-library/blob/c2922d30f2122c285b4a75db3b8b114b6e233350/src/Operation/Aggregate.php) conditionally returned non-Cursor Iterator implementations crafted from inline responses. That is certainly no longer relevant.

File: src/BulkWriteResult.php
Comment (jmikola) [reply src/BulkWriteResult.php:43]: Do you want to create a separate PHPLIB ticket related to [PHPLIB-954](https://jira.mongodb.org/browse/PHPLIB-954) and depending on [PHPC-2144](https://jira.mongodb.org/browse/PHPC-2144) to track this?

File: src/ChangeStream.php
Comment (jmikola) [inline src/ChangeStream.php]: This is temporary, correct? I assume this changes to `getCursorId(): Int64` once [PHPC-2351](https://jira.mongodb.org/browse/PHPC-2351) is implemented.

File: src/GridFS/Bucket.php
Comment (jmikola) [inline src/GridFS/Bucket.php:187]: Noted that we cannot `resource` return type hints, so the `@return resource` annotations remain in place for those methods that open a stream.

File: src/InsertManyResult.php
Comment (jmikola) [inline src/InsertManyResult.php:43]: I assume https://github.com/mongodb/mongo-php-library/pull/1391/files#r1758279785 applies here as well, yes? Along with any other result classes that proxy a WriteResult?

File: src/Operation/DatabaseCommand.php
Comment (jmikola) [inline src/Operation/DatabaseCommand.php]: Should this be CursorInterface, or is the narrow return type sensible given we know exactly what the method does? I saw @GromNaN pointed something similar out elsewhere, so I'm not sure if it's the same issue (in which case there could also be others that were missed).

File: src/Operation/Executable.php
Comment (jmikola) [inline src/Operation/Executable.php]: Consider leaving this untyped or removing the interface entirely (that should warrant a PHPLIB ticket for our internal tracking).

File: src/Operation/ListCollections.php
Comment (jmikola) [inline src/Operation/ListCollections.php:76]: I started a [Slack thread](https://mongodb.slack.com/archives/C026UA4GF0B/p1726247537366849) to discuss whether these iterators and the corresponding model classes provide any value. Even if we keep the model classes, I suppose we could remove the iterators and rely on Psalm to type them as an iterator of a particular model class.

File: src/ChangeStream.php
Comment (alcaeus) [reply src/ChangeStream.php]: That assumption is correct 👍

File: src/InsertManyResult.php
Comment (alcaeus) [reply src/InsertManyResult.php:43]: Correct

File: src/BulkWriteResult.php
Comment (alcaeus) [reply src/BulkWriteResult.php:43]: Tracking in PHPLIB-1518.

File: src/Collection.php
Comment (alcaeus) [reply src/Collection.php]: Sounds good. Added PHPC-2447 to track this.

I'll note that we shouldn't implement `IteratorAggregate`, as that would suggest one can obtain multiple iterators from a single cursor. As a cursor can't be rewound and may only ever yield a single iterator, I don't think we should consider implementing `IteratorAggregate`.

File: src/Operation/DropSearchIndex.php
Comment (alcaeus) [reply src/Operation/DropSearchIndex.php]: Sounds good. I've created PHPLIB-1519 to track the deprecation and PHPLIB-1520 for the removal of the `Executable` interface. In the meantime, I've removed the native type from `Executable::execute` in favour of a `mixed|void` type hint in the docblock. The `execute` methods of operations that would previously return `void` have been changed back to that type instead of trying to returning something semi-useful.

File: src/Operation/Explainable.php
Comment (alcaeus) [reply src/Operation/Explainable.php:31]: In that case, I'd rather keep it as `array` for the time being - no need to add a less than useful BC break for the sake of adding one.

File: src/Operation/UpdateSearchIndex.php
Comment (alcaeus) [reply src/Operation/UpdateSearchIndex.php]: This was changed back to `void`.

File: src/Operation/DatabaseCommand.php
Comment (alcaeus) [reply src/Operation/DatabaseCommand.php]: While we technically know what return type is coming from the extension, using `CursorInterface` allows us to change the actual cursor implementation without causing a BC break. I've changed this and other types to `CursorInterface&Iterator`, which we can then replace with `CursorInterface` after PHPC-2447.

File: src/Operation/Executable.php
Comment (alcaeus) [reply src/Operation/Executable.php]: Removed for the moment and added tracking tickets.

File: src/Operation/ListCollections.php
Comment (jmikola) [reply src/Operation/ListCollections.php:76]: Noted the following JIRA tickets were created:

 * [PHPLIB-1522](https://jira.mongodb.org/browse/PHPLIB-1522) to deprecate the iterators and interfaces in 1.x
 *  * [PHPLIB-1523](https://jira.mongodb.org/browse/PHPLIB-1523) to remove them in 2.0 (replaced with templated iterator return types)

We agreed to leave the model classes in place.

File: src/Model/CodecCursor.php
Comment (jmikola) [inline src/Model/CodecCursor.php]: Offhand, would Psalm actually complain about a redundant type hint once its stubs are updated for CursorInterface extending Iterator?

File: src/Operation/Find.php
Comment (jmikola) [inline src/Operation/Find.php]: `$cursor` is assigned a few lines above, so I'm curious why phpcs is unable to find it. Should this warning be suppressed?

File: tests/Database/DatabaseFunctionalTest.php
Comment (jmikola) [inline tests/Database/DatabaseFunctionalTest.php]: Noted we're not bothering with additional checks for `Iterator::class` since it'd just become irrelevant after CursorInterface extends it in 2.0.

File: tests/Operation/WatchFunctionalTest.php
Comment (jmikola) [inline tests/Operation/WatchFunctionalTest.php]: I'm not sure why this was using `assert()` instead of a PHPUnit assertion. Apart from this, I only see `assert()` used in a couple of spec prose tests and TestFileCodec.

Consider replacing with `self::assertInstanceOf(CursorInterface::class, $cursor)` if you like.

**Review by jmikola — DISMISSED** [review]
Optional suggestions.

Decide how to handle the CI warnings, but LGTM.

**Comment by alcaeus** [comment]
Updated the PR. This currently includes changes from #1427, but I'll wait until that PR has been merged up to 2.x and rebasing this PR to preserve proper history.

File: src/Collection.php
Comment (alcaeus) [reply src/Collection.php]: `CursorInterface` was changed to extend `Iterator`, and this PR now includes changes for PHPLIB-1114, using `CursorInterface` where possible.

File: src/Operation/DatabaseCommand.php
Comment (alcaeus) [reply src/Operation/DatabaseCommand.php]: This was changed to `CursorInterface`

File: src/Operation/Find.php
Comment (alcaeus) [reply src/Operation/Find.php]: This has now been solved.

File: src/Model/CodecCursor.php
Comment (alcaeus) [reply src/Model/CodecCursor.php]: Not sure - this went out the door with PHPC-2447 being implemented before I got done with these changes.

File: src/Operation/Watch.php
Comment (jmikola) [inline src/Operation/Watch.php:357]: Nothing relevant to this PR, but I want to point out that it's a bit awkward that we can assign subscribers to Managers but not individual Servers. That forces us to use a global subscriber here.

I suppose we could add `Server::getManager()` to work around that :man_shrugging:

File: tests/Operation/WatchFunctionalTest.php
Comment (jmikola) [inline tests/Operation/WatchFunctionalTest.php]: I don't think `assertNotSame()` was ever correct here, as it previously would have been comparing two CursorId objects. I suggest switching this to `assertNotEquals()`.

Also, please double check any other cursor ID comparisons if it's not too much trouble.

The original line here seems to date back to https://github.com/mongodb/mongo-php-library/commit/aac3c6bd84ed39708c156d72dcfde7e97212254b

Comment (alcaeus) [reply tests/Operation/WatchFunctionalTest.php]: Ah, excellent point. I've looked through all usages of `CursorInterface::getId` and `ChangeStream::getCursorId` and confirmed that they only use `assertNotEquals` as well as loose comparisons (`==`) throughout. 👍

File: tests/Operation/WatchFunctionalTest.php
Comment (GromNaN) [inline tests/Operation/WatchFunctionalTest.php:724]: Would it make the test more robust if you don't rely on int cast?
```suggestion
        $this->assertNotEquals(new Int64(0), $changeStream->getCursorId());
```

Comment (alcaeus) [reply tests/Operation/WatchFunctionalTest.php:724]: It doesn't make a difference, as the custom `Int64Comparator` will always handle these assertions, and it leverages the comparison behaviour implemented in `Int64` directly.

File: tests/Operation/WatchFunctionalTest.php
Comment (alcaeus) [inline tests/Operation/WatchFunctionalTest.php:1375]: Note: this method previously created the fail point on the primary server, instead of on the secondary server where the changestream was created.  This meant that the change stream was never resumed and the cursor ID was the same as before. We just didn't catch this mistake as the `assertNotSame` comparison always saw the objects as different as they didn't have the same identity. Lessons learned...

Comment (GromNaN) [reply tests/Operation/WatchFunctionalTest.php:1375]: It should be fixed on the 1.x branch, right?

Comment (alcaeus) [reply tests/Operation/WatchFunctionalTest.php:1375]: Extracted to #1434. Will rebase when that PR is merged.

**Review by jmikola — APPROVED** [review]

**Review by GromNaN — APPROVED** [review]

**Pull request merged**
